### ISSUE TITLE
Sign-in popup looks like parent page for Safari

### ISSUE THREAD

**Issue by LindsayRidgeway** [MrSwitch/hello.js#181]
Hi, Drew. I'm thrilled with hello.js. It's working great in our app for obtaining email addresses from Google and Facebook for users on Chrome and Firefox, at least the latest versions running on Mac OS X. I also greatly appreciated your support a few days ago in understanding how to transfer the access token from the client to the server in order to defeat client spoofing. Our app's sign-in/sign-out/session-management is now working really nicely.

However, the sign-in process doesn't work at all with the Safari browser. When you click links that should bring up the interface to the third-party app (Google or Facebook), instead you see a replica of the original page, squeezed into a popup window.

I don't know if this is a general problem that everyone who uses "hello.js" with Safari encounters, or something unusual about our app. I'll assist in any way I can with helping to diagnose and correct the problem.

I don't know how this issues tracker works. Hopefully I'll receive an email if you respond. Just in case, I'm LDRidgeway at gmail. 

Thanks!

Lindsay

**Comment by MrSwitch** [comment]
@LindsayRidgeway 

Safari has an outstanding issue whereby it would crop the HashFragment from the URL when the an HTTP redirected was triggered to a page it hadn't been to before. Hence why in Safari you see it go to the location of 'redirect_uri' before being rerouted on to the correct location.

To keep this clean, you should define a separate `redirect_uri` page. The default is to use the current page. You can change this setting globally like so.

```javascript
hello.init(client_ids, {redirect_uri:'/redirect.html'})
```

Note: you'll also need to register the full path with the thirdparty providers.

I hope that clears it up for you, please close this ticket if you dont have any further comments.

**Comment by LindsayRidgeway** [comment]
Hi, Drew. Thanks so much for your reply, but I don't understand what you are suggesting so that it will also work with Safari.

Here's what I have now:

```javascript
    <script>
        function facebook_signin() {
            hello.init(
                    {facebook: "{{facebook_app_id}}"},
                    {redirect_uri: '/oauth_redirect'}
            );
            hello("facebook").login({scope: "email"}).then(function () {
                access_token = hello("facebook").getAuthResponse().access_token;
                ...send the token to the server
            });
        }
        function google_signin() {
            hello.init(
                    {google: "{{google_app_id}}"},
                    {redirect_uri: '/oauth_redirect'}
            );
            hello("google").login({scope: "email"}).then(function () {
                access_token = hello("google").getAuthResponse().access_token;
                ...send the token to the server
            });
        }
    </script>
```

Here is the file that is served as "/oauth_redirect" (a single line):

```javascript
<script src="/ui/scripts/hello.min.js"></script>
```

The `oauth_redirect` URLs are registered with the third party apps for all instances of the app.

Are you suggesting a change that would continue to work for Chrome and Firefox but would make Safari start working? Sorry, I'm missing what that change would be.

**Comment by MrSwitch** [comment]
What you have there should work (although   hello.init is just a config, no need to trigger it via an event)

Can you replicate the issue on any of the http://adodson.com/hello.js/demos

**Comment by LindsayRidgeway** [comment]
First, thanks for note about config. I hope this looks right now:

```javascript
        hello.init(
                {
                    facebook: "{{facebook_app_id}}",
                    google: "{{google_app_id}}"
                },
                {redirect_uri: '/oauth_redirect'}
        );
        function facebook_signin() {
            hello("facebook").login({scope: "email"}).then(function () {
                access_token = hello("facebook").getAuthResponse().access_token;
                ...send the token to the server
            });
        }
        function google_signin() {
            hello("google").login({scope: "email"}).then(function () {
                access_token = hello("google").getAuthResponse().access_token;
                ...send the token to the server
            });
        }
```

Second, good news, your login demos seems to work fine for all three third-party apps on Safari. Behavior for my code, however, is still the same.

Any changes you can suggest, anything I can try?

**Comment by MrSwitch** [comment]
Can you publish your broken implementation?

Here's another suggestion of how you might like to implement your code.

```javascript
hello.init(
    {
        facebook: "{{facebook_app_id}}",
        google: "{{google_app_id}}"
    },
    {
        redirect_uri: '/oauth_redirect',
        scope : 'email'
    }
);
function login( network ) {
    hello( network ).login().then(function (auth) {
        var access_token = auth.authResponse.access_token;
        ...send the token to the server
    });
}
```

**Comment by LindsayRidgeway** [comment]
First, I have updated my JavaScript code for interfacing with "hello.js" once again per your latest suggestion. As before, this works on Chrome and Firefox, but continues to have the behavior described above for Safari.

I'm not sure what code you want to see. I'll try to provide everything relevant.

To start, this is a GAE application, and we use Jinja2 as the template language. So here is the base template that acts as the parent for other pages, the content of which is inserted at the `{%block workspace%}{%endblock%}` Jinja2 tags. The program tests the Jinja2 variable `email` to determine whether the user is currently signed in.

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>...page-title...</title>
    <link type="text/css" rel="stylesheet" href="/ui/{{ui}}/stylesheets/main.css"/>
    <link rel="icon" href="/favicon.ico">
    {%if not email%}
    <script src="/ui/{{ui}}/scripts/hello.min.js"></script>
    <script src="//code.jquery.com/jquery-1.11.1.min.js"></script>
    <script>
        hello.init(
                {
                    facebook: "{{facebook_app_id}}",
                    google: "{{google_app_id}}"
                },
                {
                    redirect_uri: '/oauth_redirect',
                    scope: 'email'
                }
        );
        function signin(network) {
            hello(network).login().then(function (auth) {
                var access_token = auth.authResponse.access_token;
                $("#oauth_app").val(network);
                $("#oauth_token").val(access_token);
                {%if not continue_uri%}
                $("#oauth_uri").val("" + location);
                {%endif%}
                $("#oauth_form").submit();
            });
        }
    </script>
    {%endif%}
</head>
<body>
<form id="oauth_form" method="post" action="/oauth_token">
    <input id="oauth_app" type="hidden" name="app">
    <input id="oauth_token" type="hidden" name="access_token">
    {%if continue_uri%}
    <input id="oauth_uri" type="hidden" name="uri" value="{{continue_uri}}">
    {%else%}
    <input id="oauth_uri" type="hidden" name="uri">
    {%endif%}
</form>
(...page-content...)
        {%if email%}
            <a href="/signout">Sign out</a>
        {%else%}
            <i>Sign in with</i>
            <a href="javascript:signin('facebook');">Facebook</a>
            <i>or</i>
            <a href="javascript:signin('google');">Google</a>
        {%endif%}
(...page-content...)
            {%block workspace%}{%endblock%}
(...page-content...)
</body>
</html>
```

And then here is the relevant server-side Python code:

```Python
class OauthRedirect(CustomRequestHandler):
    def get(self):
        jinja_template = ui.get_template(self, u'oauth_redirect.html')
        jinja_template_values = get_standard_site_values(self.session)
        self.response.out.write(jinja_template.render(jinja_template_values))


class OauthToken(CustomRequestHandler):
    def post(self):
        app = self.request.get(u'app')
        access_token = self.request.get(u'access_token')
        uri = self.request.get(u'uri')
        email = None
        if app == u'facebook':
            profile = urlfetch.fetch(u'https://graph.facebook.com/me?access_token={}'.format(access_token))
            obj = json.loads(profile.content)
            email = obj[u'email']
        elif app == u'google':
            profile = urlfetch.fetch(u'https://www.googleapis.com/plus/v1/people/me?access_token={}'.
                                     format(access_token))
            obj = json.loads(profile.content)
            email = obj[u'emails'][0][u'value']
        self.session[u'email'] = email.lower()
        self.redirect(str(uri))

class SignIn(CustomRequestHandler):
    def get(self):
        jinja_template = ui.get_template(self, u'signin.html')
        jinja_template_values = get_standard_site_values(self.session)
        continue_uri = self.request.get(u'continue_uri')
        jinja_template_values[u'continue_uri'] = continue_uri if continue_uri else self.request.get(u'uri')
        self.response.out.write(jinja_template.render(jinja_template_values))

class SignOut(CustomRequestHandler):
    def get(self):
        self.session[u'email'] = None
        self.redirect(u'/')

# Start the application
config = {}
config['webapp2_extras.sessions'] = {
    'secret_key': '...secret-string...',
    }
app = webapp2.WSGIApplication([('/', MainPage)], config=config, debug=True)

app.router.add((u'/oauth_redirect', OauthRedirect))
app.router.add((u'/oauth_token', OauthToken))
app.router.add((u'/signin', SignIn))
app.router.add((u'/signout', SignOut))
```

Here's `signin.html`:

```html
{%extends "site_base.html"%}
{%block workspace%}
<h3>Sign In</h3>
<p>
    Please sign in by clicking one of the sign-in links above.
</p>
{%endblock%}
```

Here again is `oauth_redirect.html`:

```html
<script src="/ui/{{ui}}/scripts/hello.min.js"></script>
```

And in case you want to look at our custom request handler:

```python
from urlparse import urlparse

import webapp2
from webapp2_extras import sessions

import dao


class CustomRequestHandler(webapp2.RequestHandler):
    def dispatch(self):
        try:
            # Get the session store
            self.session_store = sessions.get_store(request=self.request)
            if u'oauth_token' in self.request.uri:
                # If this is the POST transaction from "site_base.html" delivering the OAUTH token,
                # dispatch this request
                super(CustomRequestHandler, self).dispatch()
            elif self.session and self.session[u'email']:
                if dao.test_current_user_registered(self.session):
                    # If the user is signed in and registered, dispatch this request
                    super(CustomRequestHandler, self).dispatch()
                else:
                    if dao.get_site().enable_register:
                        # If user is signed in and not registered, but we're accepting new users,
                        # register the new user, then dispatch this request
                        dao.register_current_user(self.session)
                        super(CustomRequestHandler, self).dispatch()
                    else:
                        # If the user is signed in and not registered, but we're not accepting new users,
                        # produce the "Sorry" display of the home page
                        path = urlparse(self.request.uri).path
                        if path == u'/':
                            # If this dispatch request is for "/", dispatch the page, which will
                            # produce the "Sorry" display
                            super(CustomRequestHandler, self).dispatch()
                        else:
                            # If this dispatch request is not for "/", redirect to "/", and that will
                            # then produce the "Sorry" display
                            self.redirect(u'/')
            else:
                path = urlparse(self.request.uri).path
                if path == u'/' or path == u'/signin' or path == u'/help':
                    # If the user is not signed in, and the path is one of those that the user is permitted
                    # to see without signing in, dispatch this request
                    super(CustomRequestHandler, self).dispatch()
                else:
                    # If the user is not signed in, redirect to the "signin" page, passing along the
                    # requested URI so that we can redirect to that URI after successful sign-in
                    self.redirect(u'/signin?continue_uri={}'.format(self.request.uri))
        finally:
            # Save the session store
            self.session_store.save_sessions(self.response)

    @webapp2.cached_property
    def session(self):
        # For any reference to "self.session", get the session from the session store
        return self.session_store.get_session()
```

**Comment by LindsayRidgeway** [comment]
Ah, I've fixed it! 

I put a print statement at the entry to the custom request handler and discovered that the code was dispatching an extra page. So I changed the first "if" statement to:

```python
if u'oauth_token' in self.request.uri or u'oauth_redirect' in self.request.uri:
```

The print statement shows that for some reason, even now Safari dispatches to "oauth_redirect" twice per sign-in, whereas the others only do it once. I don't know why Safari is different, but this code change took care of it.

Thanks so much for helping me with this!
